expirationwatch/expiration_watcher.go

```diff
@@ -0,0 +1,162 @@
+	key := rbt.Int64Key(expirationTimeSeconds)
+	w.rbTreeMu.Lock()
+	defer w.rbTreeMu.Unlock()
+	value, ok := w.rbTree.Get(&key)
+	if !ok {
+		w.rbTree.Insert(&key, map[common.Hash]struct{}{hash: struct{}{}})
+	} else {
+		hashes := value.(map[common.Hash]struct{})
+		hashes[hash] = struct{}{}
+		w.rbTree.Insert(&key, hashes)
+	}
+}
+
+// Remove removes the item with a specified hash from the expiration watcher
+func (w *Watcher) Remove(expirationTimeSeconds int64, hash common.Hash) {
+	key := rbt.Int64Key(expirationTimeSeconds)
+	w.rbTreeMu.Lock()
+	defer w.rbTreeMu.Unlock()
+	value, ok := w.rbTree.Get(&key)
+	if !ok {
+		// Due to the asynchronous nature of the Watcher and OrderWatcher, there are
+		// race-conditions where we try to remove an item from the Watcher after it
+		// has already been removed.
+		log.WithFields(log.Fields{
+			"hash": hash,
+		}).Trace("Attempted to remove item from Watcher that no longer exists")
+		return // Noop
+	} else {
+		hashes := value.(map[common.Hash]struct{})
+		delete(hashes, hash)
+		if len(hashes) == 0 {
+			w.rbTree.Delete(&key)
+		} else {
+			w.rbTree.Insert(&key, hashes)
+		}
+	}
+}
+
+// Start starts the expiration watchers poller
+func (w *Watcher) Start(pollingInterval time.Duration) error {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+	if w.isWatching {
+		return errors.New("Expiration watcher already started")
+	}
+	if w.wasStartedOnce {
+		return errors.New("Can only start Watcher once per instance")
+	}
+	w.isWatching = true
+	w.wasStartedOnce = true
+
+	go func() {
+		// TODO(fabio): Optimize this poller. We could keep track of soonestExpirationTime as a property of
+		// Watcher. Whenever a new item is added via Add, we check if the expiration time is sooner
+		// than soonestExpirationTime and if so, we update soonestExpirationTime. Then instead of running the
+		// inner for loop at a constant frequency, we adjust the frequency based on the value of
+		// soonestExpirationTime (probably by using time.After or time.Sleep).
+		ticker := time.NewTicker(pollingInterval)
+		for {
+			<-ticker.C
+
+			w.mu.Lock()
+			if !w.isWatching {
+				ticker.Stop()
+				close(w.expiredOrders)
+				w.mu.Unlock()
+				return
+			}
+			w.mu.Unlock()
+
+			w.expiredOrders <- w.prune()
+		}
+	}()
+	return nil
+}
+
+// Stop stops the expiration watchers poller
+func (w *Watcher) Stop() {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+	w.isWatching = false
+}
+
+// Receive returns a read-only channel that can be used to listen for expired items
+func (w *Watcher) Receive() <-chan []ExpiredItem {
+	return w.expiredOrders
+}
+
+// prune checks for any expired items, removes them from the expiration watcher and returns them
+// to the caller
+func (w *Watcher) prune() []ExpiredItem {
+	pruned := []ExpiredItem{}
+	for {
+		w.rbTreeMu.RLock()
+		key, value := w.rbTree.Min()
+		w.rbTreeMu.RUnlock()
+		if key == nil {
+			break
+		}
+		expirationTimeSeconds := reflect.ValueOf(key).Elem().Int()
```

Comment:
The reflect package is mostly for when you don't know types ahead of time, but I don't think we need it here. We know that `key` is of type `*rbt.Int64Key` so you can just do a type cast. I think it would be something like:

```
expirationTimeSeconds := *(*rbt.Int64Key)(key)
```

Reply:
@fabioberger did you mean to resolve this? The code is still the same.

Reply:
Yes, thanks for pointing this out. I've made this change.